### PR TITLE
Return a dataclass from `geqdsk.read`

### LINKED ISSUE #17: Use class for data instead of/as well as `dict`
This would give us better type hints, as well as be able to give synonyms to some of the fields as some existing readers use different names.

### PR DESCRIPTION
**Pull request by ZedThree**
Closes #17

### CODE REVIEW

File: freeqdsk/geqdsk.py [hunk 1]
```diff
@@ -182,6 +183,163 @@
...
+                r[i, j] = self.rleft + self.rdim * i / (self.nx - 1)
+                z[i, j] = (self.zmid - 0.5 * self.zdim) + self.zdim * j / (self.ny - 1)
+
+        self.r_grid = r
+        self.z_grid = z
+
+    nw = _synonym("nx")
+    nr = _synonym("nx")
+    nh = _synonym("ny")
+    nz = _synonym("ny")
```
Comment (LiamPattinson) [inline freeqdsk/geqdsk.py:319]: This is a cool feature!

File: freeqdsk/geqdsk.py [hunk 2]
```diff
@@ -182,6 +183,163 @@
...
+    zlim: FloatArray | None = None
+
+    def __post_init__(self):
+        r = np.zeros((self.nx, self.ny), float)
+        z = np.zeros((self.nx, self.ny), float)
+
+        for i in range(0, self.nx):
+            for j in range(0, self.ny):
+                r[i, j] = self.rleft + self.rdim * i / (self.nx - 1)
+                z[i, j] = (self.zmid - 0.5 * self.zdim) + self.zdim * j / (self.ny - 1)
```
Comment (LiamPattinson) [inline freeqdsk/geqdsk.py]: I know performance isn't the top priority here, but the following should do it without the for loops:

```python
r = self.rleft + self.rdim * np.arange(self.nx) / (self.nx - 1)
z = (self.zmin - 0.5 * self.zdim) + self.zdim * np.arange(self.ny) / (self.ny - 1)
self.r_grid, self.z_grid = np.meshgrid(r, z, indexing="ij")
```

File: freeqdsk/geqdsk.py [hunk 3]
```diff
@@ -182,6 +183,163 @@
...
+    ffprime: FloatArray
+    pprime: FloatArray
+    psi: FloatArray
+    qpsi: FloatArray
+    nbdry: int
+    nlim: int
+    rbdry: FloatArray | None = None
+    zbdry: FloatArray | None = None
+    rlim: FloatArray | None = None
+    zlim: FloatArray | None = None
```
Comment (LiamPattinson) [inline freeqdsk/geqdsk.py]: We'll have to set `requires-python = ">=3.10"` in `pyproject.toml` to use this notation

File: freeqdsk/geqdsk.py [hunk 4]
```diff
@@ -182,6 +183,163 @@
...
+    ylim = _synonym("zlim")
+    nbbbs = _synonym("nbdry")
+    rbbbs = _synonym("rbdry")
+    zbbbs = _synonym("zbdry")
+
+    def __getitem__(self, name: str):
+        return getattr(self, name)
+
+    def __setitem__(self, name: str, value):
+        return setattr(self, name, any)
```
Comment (LiamPattinson) [inline freeqdsk/geqdsk.py]: Should be `setattr(self, name, value)`

**Review by LiamPattinson — CHANGES_REQUESTED** [review]
I found a few minor issues, but generally I think it's a big improvement. The `__setitem__` bug and minimum Python version need addressing, but otherwise looks good.

File: freeqdsk/geqdsk.py [hunk 5]
```diff
@@ -182,6 +183,163 @@
 )
 
 
+def _synonym(canonical: str) -> property:
+    """Set a property which is just a synonym for another attribute"""
+
+    return property(
+        lambda self: getattr(self, canonical),
+        lambda self, value: setattr(self, canonical, value),
```
Comment (LiamPattinson) [inline freeqdsk/geqdsk.py:191]: Could also set `doc=f"Synonym for {canonical}"`

File: freeqdsk/geqdsk.py (hunk 2 above)
Comment (ZedThree) [reply freeqdsk/geqdsk.py]: Thanks! This was just copy-pasted from another geqdsk reader

File: freeqdsk/geqdsk.py (hunk 3 above)
Comment (ZedThree) [reply freeqdsk/geqdsk.py]: Ah yes, I might just switch back to `Optional` for now then, although SPEC0 says it's actually almost time to drop support for 3.10 even!

File: freeqdsk/geqdsk.py (hunk 5 above)
Comment (ZedThree) [reply freeqdsk/geqdsk.py:191]: Nice, great idea!

File: freeqdsk/geqdsk.py (hunk 3 above)
Comment (LiamPattinson) [reply freeqdsk/geqdsk.py]: I'm counting the days until we can finally drop 3.9. It shouldn't be hard to fix, as `ruff check --select=UP --fix` will do it all in one go.

**Review by LiamPattinson — APPROVED** [review]

**Pull request merged**